src/topology/metric_space/infsep.lean

```diff
@@ -0,0 +1,464 @@
+/-
+Copyright (c) 2022 Wrenna Robson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wrenna Robson
+-/
+import topology.metric_space.basic
+
+/-!
+# Infimum separation
+
+This file defines the extended infimum separation of a set. This is approximately dual to the
+diameter of a set, but where the extended diameter of a set is the supremum of the extended distance
+between elements of the set, the extended infimum separation is the infimum of the (extended)
+distance between *distinct* elements in the set.
+
+We also define the infimum separation as the cast of the extended infimum separation to the reals.
+This is the infimum of the distance between distinct elements of the set when in a pseudometric
+space.
+
+All lemmas and definitions are in the `set` namespace to give access to dot notation.
+
+## Main definitions
+* `set.einfsep`: Extended infimum separation of a set.
+* `set.infsep`: Infimum separation of a set (when in a pseudometric space).
+
+!-/
+
+variables {α β : Type*}
+
+namespace set
+
+section einfsep
+open_locale ennreal
+open function
+
+/-- The "extended infimum separation" of a set with an edist function. -/
+noncomputable def einfsep [has_edist α] (s : set α) : ℝ≥0∞ :=
+⨅ (x ∈ s) (y ∈ s) (hxy : x ≠ y), edist x y
+
+section has_edist
+variables [has_edist α] {x y : α} {s t : set α}
+
+lemma le_einfsep_iff {d} : d ≤ s.einfsep ↔ ∀ (x y ∈ s) (hxy : x ≠ y), d ≤ edist x y :=
+by simp_rw [einfsep, le_infi_iff]
```

Comment:
We have a tool to express precisely properties that are satisfied by pairs of distinct elements in a set. It is called `set.pairwise` (and comes with some API). A way to reformulate your lemma using it would be:
```suggestion
lemma le_einfsep_iff {d} : d ≤ s.einfsep ↔ s.pairwise (λ x y, d ≤ edist x y) :=
by simp_rw [einfsep, le_infi_iff, set.pairwise]
```
It would probably be more mathlib-idiomatic to express lemmas like this when possible, but I will understand if you prefer to stick with the current formulation. All the lemmas where you see `∀ (x y ∈ s) (hxy : x ≠ y)` are probably eligible to this.

Reply:
I will experiment with it and see how I find it.

Reply:
What's the exists-equivalent?